functions/src/error-handling.ts

```diff
@@ -0,0 +1,55 @@
+import { randomUUID } from "crypto";
+import { Response } from "firebase-functions";
+import { ResponseType } from "./utils";
+
+const SHARE_LINK_ERRORS = {
+  400: "Invalid Request.",
+  404: "Not Found.",
+  500: "Internal API Failure.",
+};
+type ShareLinkErrorCode = keyof typeof SHARE_LINK_ERRORS;
+
+/**
+ * Custom error handling class for share link API errors to log errors internally
+ * and send a user-friendly error message to the client.
+ *
+ * @param code HTTP status code for the error.
+ * @param response Firebase functions response to send external error to.
+ * @param external Error message to send to the client. Message should not include error code.
+ * @param internal Error or error message to log to the console. If not provided, externalMessage is used.
+ * @param responseType Response type to send to the client. Defaults to text/html.
+ */
+export class ShareLinkError extends Error {
+  constructor(
+    errorCode: ShareLinkErrorCode,
+    response: Response<unknown>,
+    external: string,
+    internal?: string | Error,
+    responseType: ResponseType = ResponseType.TEXT
+  ) {
+    const errorId = randomUUID();
+    const errorIdText = `If this continues, please reach out to us and reference error ID ${errorId}.`;
+
+    // Only send the errorId to the client if it's an internal server error.
+    const isServerError = Math.floor(errorCode / 100) * 100 === 500;
+    const externalMessage =
+      `Error ${errorCode}: ${SHARE_LINK_ERRORS[errorCode]} ` +
+      `${external} ${isServerError ? errorIdText : ""}`;
+
+    response
+      .status(errorCode)
+      .send(formatMessage(externalMessage, responseType));
```

Comment:
Yeah, I agree that shoehorning the response sending into the exception handling itself probably isn't the best way to approach this. 

I've reworked this again using your `SHARE_LINK_ERRORS`/`ShareLinkErrorCode` suggestions from below, and I'm using the `ShareLinkError` class strictly as a wrapper of the default `Error` class that corresponds to a specific `ShareLinkErrorCode`. 

This way we can explicitly throw `ShareLinkErrors()` for errors that we're aware of so that we can identify them (similar to what you suggested above), and otherwise throw a generic 500/internal error. 

Something like:
```ts
.catch((error: Error) => {
  if (error instanceof ShareLinkError) {
      res.status(error.httpCode).send(error.message);
      throw error;
  } else {
    const unexpectedError = getShareLinkErrorByCode(ShareLinkErrorCode.UNEXPECTED_ERROR)
    res.status(unexpectedError.httpCode).send(unexpectedError.message); // send generic 500 error
    throw error // log the original error internally
  }
}
```